### PR TITLE
kv/RocksDBStore: Table options for indexing and filtering

### PR DESCRIPTION
**Pull request by markhpc**
This PR exposes a number of table options for RocksDB backed key/value stores related to filtering and indexing.  It also enables bloom filters by default with 20 bits per key.  In testing, this resulted in a significant performance increase (possibly orders of magnitude) when writing out new objects to HDD vs no bloom filters, and about a 20% increase in performance vs setting 10 bits per key.  The other newly exposed options may be valuable on clusters where the k/v store has grown big enough that indexes/filters can no longer fit into cache all at once, but the partition index filters can not yet be used since they are still experimental.

### CODE REVIEW

**Comment by markhpc** [comment]
Note to self: Update doc/rados/configuration/osd-config-ref.rst

File: src/common/config_opts.h [hunk 1]
```diff
@@ -1217,8 +1217,16 @@ OPTION(bluestore_debug_random_read_err, OPT_DOUBLE, 0)
...
+
 OPTION(kstore_rocksdb_options, OPT_STR, "compression=kNoCompression")
-OPTION(kstore_rocksdb_bloom_bits_per_key, OPT_INT, 0)
+OPTION(kstore_rocksdb_bloom_bits_per_key, OPT_INT, 20)
+OPTION(kstore_rocksdb_cache_index_and_filter_blocks, OPT_BOOL, false)
+OPTION(kstore_rocksdb_cache_index_and_filter_blocks_with_high_priority, OPT_BOOL, false)
+OPTION(kstore_rocksdb_index_type, OPT_STR, "binary_search")
+OPTION(kstore_rocksdb_partition_filters, OPT_BOOL, false)
+OPTION(kstore_rocksdb_metadata_block_size, OPT_U64, 0)
+OPTION(kstore_rocksdb_pin_l0_filter_and_index_blocks_in_cache, OPT_BOOL, false)
```
Comment (liewegas) [inline src/common/config_opts.h]: everything from kstore_rocksdb_options should just be rocksdb_*.. none of this is related in any way to os/KStore (except that kstore *might* use rocksdb as its backend)

Comment (markhpc) [inline src/common/config_opts.h]: So we are already using kstore_bloom_bits_per_key to set the bloom bits generally for rocksdb in kv/RocksDBStore.cc.  That's why I used the kstore prefix to match what we were already doing.  Should we change that as well?

**Comment by liewegas** [comment]
Yeah, i think that was just a mistake :)  thanks!

File: src/common/config_opts.h (hunk 1 above)
Comment (liu-chunmei) [inline src/common/config_opts.h]: will the other BlockBasedTableOptions structure members influence the performance? Do we need set one by one?

**Comment by liupan1111** [comment]
@markhpc how about the performance improvement in SSD?

**Comment by yuriw** [comment]
@markhpc 
may need rebase

From https://github.com/markhpc/ceph
 * branch            wip-rocksdb-bloomsettings -> FETCH_HEAD
CONFLICT (modify/delete): src/common/config_opts.h deleted in HEAD and modified in bc0d7fe0ff945c202679e1b35911130276eea1ee. Version bc0d7fe0ff945c202679e1b35911130276eea1ee of src/common/config_opts.h left in tree.
Automatic merge failed; fix conflicts and then commit the result.

**Comment by markhpc** [comment]
liupan: need to run new tests across the board since the partition filters are experimental since we can't use them.  Going to rerun tests now.

**Comment by markhpc** [comment]
The commits got a little messy after integrating with #16527 , so I ended up just folding everything into one as it's not that huge anyway.

**Comment by markhpc** [comment]
retest this please

**Comment by markhpc** [comment]
retest this please

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-27_18:08:03-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**